python/ray/rllib/agents/dqn/simple_q_model.py

```diff
@@ -11,29 +12,46 @@
 class SimpleQModel(TFModelV2):
     """Extension of standard TFModel to provide Q values."""
 
-    def get_action_scores(self, state_embedding, hiddens):
-        """Returns Q(s, a) given an state embedding tensor.
+    def __init__(self,
+                 obs_space,
+                 action_space,
+                 num_outputs,
+                 model_config,
+                 name,
+                 q_hiddens=(256, )):
+        super(SimpleQModel, self).__init__(obs_space, action_space,
+                                           num_outputs, model_config, name)
+
+        # setup the Q head output (i.e., model for get_q_values)
+        self.model_out = tf.keras.layers.Input(shape=(num_outputs, ),
+                                               name="model_out")
+
+        if q_hiddens:
+            last_layer = self.model_out
+            for i, n in enumerate(q_hiddens):
```

Comment:
> Hm isn't either way compatible with eager?

It is, but it does this by creating "non-eager" tensors on which it attaches a hidden trace. My main issue is that it is a regression towards some magic hidden state. This seems to do something very similar to what `tf.function` does but in a much less explicit way. Decorating a function with `tf.function` clearly labels the part of the code that is traced, while I find this syntax doesn't delineated very clearly which tensors are actually eager and which ones are not. I could see this style be more error prone once the majority of users get used to TF eager/2.0 and more complex to mix and match with non-keras code.

> I like the existing style since it seems more explicit what is going on, for example I'm not sure what is going on when you use tf.keras.Sequential, it seems magical that it can create that without knowing the input shapes.

If you provide an input layer or a layer with its input dimensions defined, I think the variables are constructed right away. In this particular example, the shapes would be inferred the same way in either implementation. However, the fully OO way allows for variable creation to be deferred until the first call at which time the shapes can be inferred before they are needed.

The logic for that isn't very complex and is the recommended keras API for [custom layers](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer). It just follows something along the lines of:
```python
def __call__(self, inp):
    if not self._built:
        self.create_variables(inp)
    self.call(inp)
```

That being said, since this is just an example, I don't see any issues with leaving it as is if that is what you prefer.